src/dust/coagulation/coagulation.hpp

```diff
@@ -0,0 +1,1000 @@
+  mbr.team_barrier();
+
+  // Adding coagulation source terms
+  parthenon::par_for_inner(DEFAULT_INNER_LOOP_PATTERN, mbr, 0, mimax, [&](const int i) {
+    for (int j = 0; j <= i; j++) {
+      // Calculate the Rate
+      const Real fett_t = CoagulationRate(i, j, kernel4, vel, stoppingTime, coag, 0);
+      const Real Rc1 = distri(i) * distri(j) * fett_t;
+      for (int nz = 0; nz < 4; nz++) {
+        const int k = coag.cpod_notzero(i, j, nz);
+        if (k < 0) continue;
+        const Real src_coag0 = coag.cpod_short(i, j, nz) * Rc1;
+        Kokkos::atomic_add(&source(k), src_coag0);
+      }
+    }
+  });
+  mbr.team_barrier();
+
+  // FRAGMENTATION------------------------------------------------------------------------
+
+  // Adding fragment distribution
+  const int pgrid = coag.pgrid;
+  int iafrag = coag2drv::afrag;
+  int iphifrag = coag2drv::phifrag, iepsfrag = coag2drv::epsfrag;
+  parthenon::par_for_inner(
+      DEFAULT_INNER_LOOP_PATTERN, mbr, 0, coag.nm - 1, [&](const int k) {
+        for (int j = k; j <= mimax; j++) {
+          // Calculate As(j) on fly
+          Real As_j = 0.0;
+          for (int i2 = 0; i2 <= mimax; i2++) {
+            for (int j2 = 0; j2 <= i2; j2++) {
+              if (coag.klf(i2, j2) == j) {
+                const Real fett_l =
+                    CoagulationRate(i2, j2, kernel4, vel, stoppingTime, coag, 1);
+                As_j += coag.coagR3D(iafrag, i2, j2) * distri(i2) * distri(j2) * fett_l;
+              }
+            }
+          }
+          source(k) += coag.coagR3D(iphifrag, k, j) / coag.mass_grid(k) * As_j;
+        }
+      });
+  mbr.team_barrier();
+
+  // Negative terms and cratering remnants
+  parthenon::par_for_inner(DEFAULT_INNER_LOOP_PATTERN, mbr, 0, mimax, [&](const int j) {
+    Real sum0 = 0.0;
+    for (int i = j; i <= mimax; i++) {
+      const Real fett_l = CoagulationRate(i, j, kernel4, vel, stoppingTime, coag, 1);
+      const Real Rf1 = distri(i) * distri(j) * fett_l;
+      sum0 -= Rf1;
+    }
+    source(j) += sum0;
+  });
+  mbr.team_barrier();
+
+  // Cratering
+  parthenon::par_for_inner(DEFAULT_INNER_LOOP_PATTERN, mbr, 0, mimax, [&](const int i) {
+    Real sum0 = 0.0;
+    for (int j = 0; j <= i - pgrid - 1; j++) {
+      const Real fett_l = CoagulationRate(i, j, kernel4, vel, stoppingTime, coag, 1);
+      const Real Rf1 = distri(i) * distri(j) * fett_l;
+      const Real dummy = coag.coagR3D(iepsfrag, i, j) * Rf1;
+      sum0 += dummy;
+    }
+
+    if (i - pgrid - 1 >= 0) {
+      Kokkos::atomic_add(&source(i - 1), sum0);
+    }
+
+    Real sum1 = -sum0;
+    // Full fragmentation (only negative terms)
+    int i1 = std::max(0, i - pgrid);
+    for (int j = i1; j <= i; j++) {
+      const Real fett_l = CoagulationRate(i, j, kernel4, vel, stoppingTime, coag, 1);
+      const Real Rf1 = distri(i) * distri(j) * fett_l;
+      sum1 -= Rf1;
+    }
+    Kokkos::atomic_add(&source(i), sum1);
+  });
```

Comment:
```suggestion
    Real sum1 = -sum0;
    // Full fragmentation (only negative terms)
    int i1 = std::max(0, i - pgrid);
    for (int j = i1; j <= i; j++) {
      const Real fett_l = CoagulationRate(i, j, kernel4, vel, stoppingTime, coag, 1);
      const Real Rf1 = distri(i) * distri(j) * fett_l;
      sum1 -= Rf1;
    }
    source(i) += sum1;
  });
```
if the loop above this is modified as suggested